File: solutions/045-jump-game-ii/solution.go
package main

import "fmt"

func jump(nums []int) int { //反向找
	n := len(nums) - 1
	step := 0
	for n > 0 {
		for i := 0; i < n; i++ {
			if i+nums[i] >= n { //找到能跳到目标的位置，然后这就是下一个目标的位置
				n = i
				step++
				break
			}
		}
	}
	return step
}
func jump2(nums []int) int { //正向找
	maxPos := 0
	end := 0
	step := 0
	for i := 0; i < len(nums)-1; i++ {
		if i+nums[i] > maxPos {
			maxPos = i + nums[i]
		}
		if i == end { //表示已经到达了当前能跳到的最远距离，需要更新下一个能跳到的最远距离
			end = maxPos
			step++
		}
	}
	return step
}

func main() {
	got := jump([]int{2, 3, 1, 1, 4})
	fmt.Printf("PASS | jumps=%d\n", got)
}

File: solutions/045-jump-game-ii/solution.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums) - 1
        step = 0
        while n > 0:
            for i in range(n):
                if i + nums[i] >= n:
                    n = i
                    step += 1
                    break
        return step

    def jump2(self, nums: List[int]) -> int:
        max_pos = 0
        end = 0
        step = 0
        for i in range(len(nums) - 1):
            if i + nums[i] > max_pos:
                max_pos = i + nums[i]
            if i == end:
                end = max_pos
                step += 1
        return step


if __name__ == "__main__":
    sol = Solution()
    got = sol.jump([2, 3, 1, 1, 4])
    status = "PASS" if got == 2 else "FAIL"
    print(f"{status} | jumps={got} (expected 2)")
